rfd/0152-automatic-database-users-mongodb.md

```diff
@@ -0,0 +1,202 @@
+---
+authors: STeve Huang (xin.huang@goteleport.com)
+state: draft
+---
+
+# RFD 152 - Database Automatic User Provisioning for MongoDB
+
+## Required approvers
+
+Engineering: @r0mant || @smallinsky
+Product: @klizhentas || @xinding33
+Security: @reedloden || @jentfoo
+
+## What
+
+This RFD discusses on how to expand Database Automatic Provisioning feature for
+MongoDB.
+
+## Why
+
+Automatic User Provisioning has been implemented for several SQL databases,
+including PostgreSQL and MySQL, with the basic design described in [RFD
+113](https://github.com/gravitational/teleport/blob/master/rfd/0113-automatic-database-users.md).
+
+Adding support for database user provisioning to MongoDB presents unique
+challenges due to differences in architecture compared to traditional SQL
+databases.
+
+This RFD aims to identify the challenges and provide solutions to address them.
+
+Since the differences in architecture between MongoDB Atlas and self-hosted
+MongoDB are significant, they will be discussed separately in this RFD.
+
+## MongoDB Atlas Details
+
+Automatic User Provisioning will NOT be supported for MongoDB Atlas with the
+reasons discussed below. This RFD should be updated if better solutions are
+found in future iterations.
+
+[Database
+Users](https://www.mongodb.com/docs/atlas/security-add-mongodb-users/) and
+[Custom Database
+Roles](https://www.mongodb.com/docs/atlas/security-add-mongodb-roles/) for
+MongoDB Atlas are managed at a Atlas project level.
+
+As a consequence, database users and roles are NOT modifiable through
+in-database connections. Instead, one can authenticate with Atlas using the
+Atlas SDK and use APIs to manage these database users and roles. Multiple
+deployment jobs will be created to update the MongoDB clusters in this project,
+upon successful APIs calls.
+
+In my personal testing on an Atlas project with a single MongoDB cluster, it
+takes 10~20 seconds for the deployment job to refresh the database user in the
+target MongoDB cluster.
+
+With the current design of Automatic User Provisioning, the database user must
+be updated with new role assignments for each new connection then the roles
+should be revoked once the connection is done. However, waiting for 10+ seconds
+for provisioning the database user each connection will result in a very bad
+user experience (also client may just time out).
+
+## Self-hosted MongoDB Details
+
+The overall flow and logic will follow the previous [RFD
+113](https://github.com/gravitational/teleport/blob/master/rfd/0113-automatic-database-users.md).
+Differences will be outlined in the sections below.
+
+### The admin user connection
+
+The Database Service will connect as an admin user in order to manage database
+users. The admin user requires the following privileges:
+```json
+{
+  "createUser": "CN=teleport-admin",
+  "roles": [
+    { "role": "clusterMonitor", "db": "admin" },
+    { "role": "userAdminAnyDatabase", "db": "admin" }
```

Comment:
This is a broad builtin role, is there a way to craft a more limited user-defined role that grants access to `createUser` and `grantRole` individually?

Reply:
Here is the "minimum" role:
```json
{
    "createRole": "teleport-admin-role",
    "privileges": [
        { "resource": { "cluster": true }, "actions": [ "grantRole", "revokeRole", "inprog" ] }, 
        { "resource": { "db": "$external", "collection": "" }, "actions": [ "createUser", "updateUser", "dropUser", "viewUser", "setAuthenticationRestriction", "changeCustomData"] },
        { "resource": { "db": "admin", "collection": "" }, "actions": [ "createRole"] }
    ],                                                                     
    "roles": [ ]
}
```

- `inprog` is for `currentOp`

Though, with `grantRole` on the cluster, you can pretty much grant yourself  anything (i mean that's the idea so `teleport-admin` can grant any role to auto-provisioned user), so the "effective" privilege `teleport-admin` has is the superuser.

Reply:
Looking at [createUser](https://www.mongodb.com/docs/manual/reference/method/db.createUser/#mongodb-method-db.createUser) documentation, it supports arbitrary "custom data" for users. Can we use that to track that they were provisioned by Teleport instead of using `teleport-admin` group and get rid of the requirement to have `createRole` permission this way?